drivers: mcux: flexio: Added generic MCUX FlexIO driver

MikhailSiomin opened the pull request:
The driver has been tested on a mimxrt1064_evk board with 4 FlexIO SPI running at the same time.

`prj.conf`:
```
CONFIG_SPI=y
# CONFIG_SPI_ASYNC=y # supported
CONFIG_SPI_MCUX_FLEXIO=y
```

`mimxrt1064_evk.overlay`:
```dts
&pinctrl {
	pinmux_flexio3spi0: pinmux_flexio3spi0 {
		group0 {
			pinmux =
				<&iomuxc_gpio_ad_b0_09_gpio1_io09>, /* cs */
				<&iomuxc_gpio_ad_b0_10_gpio1_io10>, /* cs */
				<&iomuxc_gpio_ad_b1_10_flexio3_flexio10>, /* sck */
				<&iomuxc_gpio_ad_b1_11_flexio3_flexio11>, /* sdo */
				<&iomuxc_gpio_ad_b1_04_flexio3_flexio04>; /* sdi */
			drive-strength = "r0-6";
			slew-rate = "slow";
			nxp,speed = "150-mhz";
		};
	};
	pinmux_flexio3spi1: pinmux_flexio3spi1 {
		group0 {
			pinmux =
				<&iomuxc_gpio_ad_b0_02_gpio1_io02>, /* cs */
				<&iomuxc_gpio_ad_b0_03_gpio1_io03>, /* cs */
				<&iomuxc_gpio_ad_b1_05_flexio3_flexio05>, /* sck */
				<&iomuxc_gpio_ad_b1_01_flexio3_flexio01>, /* sdo */
				<&iomuxc_gpio_ad_b1_00_flexio3_flexio00>; /* sdi */
			drive-strength = "r0-6";
			slew-rate = "slow";
			nxp,speed = "150-mhz";
		};
	};
	pinmux_flexio3spi2: pinmux_flexio3spi2 {
		group0 {
			pinmux =
				<&iomuxc_gpio_ad_b0_11_gpio1_io11>, /* cs */
				<&iomuxc_gpio_ad_b1_07_flexio3_flexio07>, /* sck */
				<&iomuxc_gpio_ad_b1_06_flexio3_flexio06>, /* sdo */
				<&iomuxc_gpio_ad_b1_08_flexio3_flexio08>; /* sdi */
			drive-strength = "r0-6";
			slew-rate = "slow";
			nxp,speed = "150-mhz";
		};
	};
	pinmux_flexio3spi3: pinmux_flexio3spi3 {
		group0 {
			pinmux =
				<&iomuxc_gpio_ad_b1_00_gpio1_io16>, /* cs */
				<&iomuxc_gpio_ad_b1_02_flexio3_flexio02>, /* sck */
				<&iomuxc_gpio_ad_b1_03_flexio3_flexio03>, /* sdo */
				<&iomuxc_gpio_ad_b1_15_flexio3_flexio15>; /* sdi */
			drive-strength = "r0-6";
			slew-rate = "slow";
			nxp,speed = "150-mhz";
		};
	};
};

&flexio3 {
	status = "okay";
	flexio3_spi0: flexio3_spi0 {
		compatible = "nxp,imx-flexio-spi";
		status = "okay";
		#address-cells = <1>;
		#size-cells = <0>;
		cs-gpios = <&gpio1 9 GPIO_ACTIVE_LOW>,
				<&gpio1 10 GPIO_ACTIVE_LOW>;
		sdo-pin = <11>;
		sdi-pin = <4>; 
		sck-pin = <10>;
		pinctrl-0 = <&pinmux_flexio3spi0>;
		pinctrl-names = "default";
		flexio3_spi0_device0: flexio-spi-device@0 {
			compatible = "vnd,spi-device";
			status = "okay";
			reg = <0>;
			spi-max-frequency = <7500000>;
			frame-format = <0>;
		};
		flexio3_spi0_device1: flexio-spi-device@1 {
			compatible = "vnd,spi-device";
			status = "disabled";
			reg = <1>;
			spi-max-frequency = <100000>;
			frame-format = <0>;
		};
	};
	flexio3_spi1: flexio3_spi1 {
		compatible = "nxp,imx-flexio-spi";
		status = "okay";
		#address-cells = <1>;
		#size-cells = <0>;
		cs-gpios = <&gpio1 2 GPIO_ACTIVE_LOW>,
				<&gpio1 3 GPIO_ACTIVE_LOW>;
		sdo-pin = <1>;
		sdi-pin = <0>; 
		sck-pin = <5>;
		pinctrl-0 = <&pinmux_flexio3spi1>;
		pinctrl-names = "default";
		flexio3_spi1_device0: flexio-spi-device@0 {
			compatible = "vnd,spi-device";
			status = "disabled";
			reg = <0>;
			spi-max-frequency = <1000000>;
			frame-format = <0>;
		};
		flexio3_spi1_device1: flexio-spi-device@1 {
			compatible = "vnd,spi-device";
			status = "okay";
			reg = <1>;
			spi-max-frequency = <7500000>;
			frame-format = <0>;
		};
	};
	flexio3_spi2: flexio3_spi2 {
		compatible = "nxp,imx-flexio-spi";
		status = "okay";
		#address-cells = <1>;
		#size-cells = <0>;
		cs-gpios = <&gpio1 11 GPIO_ACTIVE_LOW>;
		sdo-pin = <6>;
		sdi-pin = <8>; 
		sck-pin = <7>;
		pinctrl-0 = <&pinmux_flexio3spi2>;
		pinctrl-names = "default";
		flexio3_spi2_device0: flexio-spi-device@0 {
			compatible = "vnd,spi-device";
			status = "okay";
			reg = <0>;
			spi-max-frequency = <7500000>;
			frame-format = <0>;
		};
	};
	flexio3_spi3: flexio3_spi3 {
		compatible = "nxp,imx-flexio-spi";
		status = "okay";
		#address-cells = <1>;
		#size-cells = <0>;
		cs-gpios = <&gpio1 16 GPIO_ACTIVE_LOW>;
		sdo-pin = <3>;
		sdi-pin = <15>; 
		sck-pin = <2>;
		pinctrl-0 = <&pinmux_flexio3spi3>;
		pinctrl-names = "default";
		flexio3_spi3_device0: flexio-spi-device@0 {
			compatible = "vnd,spi-device";
			status = "okay";
			reg = <0>;
			spi-max-frequency = <7500000>;
			frame-format = <0>;
		};
	};
};

&lpspi3 {
	status = "disabled";
};
```

Signed-off-by: Mikhail Siomin <victorovich.01@mail.ru>

MikhailSiomin commented:
The driver will not pass the tests until the drivers of HAL are enabled (https://github.com/zephyrproject-rtos/hal_nxp/pull/205).
Also, the driver will not pass the tests if they are run on the board, because tests expect SPI dummy data to be 0x00 (https://github.com/zephyrproject-rtos/hal_nxp/pull/223).

decsny commented:
Can you please break the changes into multiple commits

GrixaYrev commented:
It is good news, FlexIO is usefull periphery. Can i help to accelerate the driver integration?

teburd commented:
@danieldegrasse @dleach02 please take a look

danieldegrasse reviewed (commented):
In general, I like the direction this PR takes to enable FlexIO. A few comments on the design of the resource manager driver:
- Can we add a way for users to register new drivers downstream without modifying the FlexIO driver in tree?
- Can interrupt management avoid iterating through all registered driver types to find the ISR function to call? I worry this could become slow if several driver types are implemented
- I would prefer we avoid defining a custom API struct for this peripheral, and just call the driver functions directly (since this is a custom API)

Also, ping @henrikbrixandersen

MikhailSiomin commented:
@danieldegrasse, thank you very much for detailed analysis.

MikhailSiomin commented on dts/bindings/misc/nxp,imx-flexio.yaml:
```diff
@@ -0,0 +1,19 @@
+
+description: NXP i.MX FlexIO controller
+
+compatible: "nxp,imx-flexio"
```
@danieldegrasse, then it might be better to rename it to `nxp,mcux-flexio` to emphasize the peripheral connection?

henrikbrixandersen replied:
No, just `nxp, 
flexio`. MCUX is a software layer.

MikhailSiomin commented on include/zephyr/dt-bindings/clock/imx_ccm.h:
```diff
@@ -55,4 +55,7 @@
 #define IMX_CCM_QTMR_CLK		0x0D00UL
 
+#define IMX_CCM_FLEXIO1_CLK		0x0E00UL
+#define IMX_CCM_FLEXIO2_3_CLK		0x0E01UL
```
@danieldegrasse, the user may mistakenly try to set the clock differently, then the last setting will be applied. The current macro will force the unknowing user to look at the documentation.

danieldegrasse replied:
Fair enough. Let's leave these constants grouped

MikhailSiomin commented on dts/bindings/spi/nxp,imx-flexio-spi.yaml:
```diff
@@ -0,0 +1,38 @@
+    type: int
+    required: true
+    description: |
+      Pin select for enable. Not really used, but need for HAL.
```
@danieldegrasse, I thought about leaving the ability to use hardware cs, but now it seems to me that it is worth using only software (cs-gpios).

MikhailSiomin commented on samples/drivers/misc/mcux_flexio/spi/src/main.c:
```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2023, STRIM, ALC
+ *
```
@danieldegrasse, the idea was to show the flexibility of setting and the ability of flexio to work in harsh conditions.

MikhailSiomin commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,481 @@
+	spi_context_cs_control(&data->ctx, true);
+
+	flexio_api->lock(config->flexio_dev);
+	flexio_api->interrupt_disable(config->flexio_dev);
```
@danieldegrasse, since the `modify-read-write` operation on the `SHIFTSIEN` register occurs in the `FLEXIO_SPI_MasterTransferNonBlocking()` function, which is called from both the user thread and the interrupt handler. Also, the `FLEXIO_SPI_MasterTransferHandleIRQ()` function changes this register from the interrupt handler.

MikhailSiomin commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,481 @@
+	spi_mcux_transfer_next_packet(data->dev);
+}
+
+#define USE_FSL_HAL_FOR_INIT (0)
```
@danieldegrasse, I will remove the initialization with hal.
Due to the fact that the `chip select` cannot be continuously active as required, the `chip select` is controlled by `software`. Therefore, it makes no sense to initialize the hardware timer under the chip select.
Another difference between the chosen initialization method is that you can adjust the clock polarity.

danieldegrasse commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -226,6 +226,53 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
+
+#if (DT_NODE_HAS_STATUS(DT_NODELABEL(flexio2), okay) \
+		 || DT_NODE_HAS_STATUS(DT_NODELABEL(flexio3), okay)) && CONFIG_MCUX_FLEXIO
+	case IMX_CCM_FLEXIO2_3_CLK:
```
Resolving, as FlexIO2 and FlexIO3 share a clock root

danieldegrasse commented on drivers/misc/mcux_flexio/mcux_flexio.c:
```diff
@@ -0,0 +1,264 @@
+};
+
+static const struct mcux_flexio_child_config children_config[NXP_IMX_FLEXIO_LAST_INDEX] = {
+	[NXP_IMX_FLEXIO_SPI] = { spi_mcux_isr, spi_mcux_res },
```
@MikhailSiomin Do you have any thoughts here? This is my core concern with the current design of this "resource management" driver. I'd like to design this code in such a way that downstream FlexIO drivers can easily register with this driver, without editing the upstream source

soburi commented on dts/bindings/misc/nxp,imx-flexio.yaml:
```diff
@@ -0,0 +1,19 @@
+
+compatible: "nxp,imx-flexio"
+
+include: base.yaml
```
Could you create yaml for define device classification?
raspi_pico has configurable-IO function that is similar to flexio.
#44316

I think it is enough with just place holder for define classification.
No need to define API because these devices used from their client drivers.

@danieldegrasse @teburd @yonsch 
Please comment if you have any opinion.
I would like to consider implementing rpi_pico to reflect this.

yonsch replied:
@soburi Not sure I understand this correctly, do you mean changing the current PIO PR? Because currently I don't see how it's different from what you're suggesting. Anyway, it should be discussed in the PIO PR.
If you meant changing the flex IO PR to be more like the PIO, then I generally agree, though I'm not really familiar with the flex IO enough to say. If possible, I do think that the generic driver should not require modifying it for every child device, but rather be something that child drivers can inherit from/incorporate into.

soburi replied:
@yonsch 
Sorry for the confusion.

I thought that if a common definition was created, it may affect PIO, so I notified it.

(If it is to be standardized, I also think it would be better to join the discuss how to make it a "preferred" form based on the knowledge of PIO.)

yonsch replied:
@soburi this will be discussed in next week's architecture WG meeting

danieldegrasse replied:
@soburi I think these devices probably do belong in "misc". I am less familiar with the PIO, but it is programmed with a set of instructions, correct? The FlexIO may be able to implement similar Zephyr driver classes, but it is not configured in the same manner. I think misc drivers make sense here, because the API to manage resources for each peripheral will be very specific to that peripheral (and may not always be required), so attempting to find similarities in the APIs to manage these peripherals does not make much sense to me. I think the primary similarity between these peripherals is that they can emulate some of the same driver classes in Zephyr.

danieldegrasse commented:
@MikhailSiomin Just wanted to add some notes from the discussion on https://github.com/zephyrproject-rtos/zephyr/pull/44316:

- It seems like the approach of defining devices implemented using FlexIO as child nodes of the FlexIO device with custom compatibles aligns well with what the PIO implementation does: https://github.com/zephyrproject-rtos/zephyr/pull/44316/files#diff-b1c2d68b9488f2b77e415a6ba1aab5e1642219584b1d756a8fa9bbe6924b1eaeR25
- The PIO driver does all resource management at runtime: https://github.com/zephyrproject-rtos/zephyr/pull/44316/files#diff-c5a3c9b64c6a358394b5a8d03d18af2db0299f030fbcbad1056077139a13c4abR18. I think this is how we should approach the FlexIO driver as well, to enable drivers to request resources from the FlexIO resource management driver when they initialize. This will also make implementing downstream drivers possible.
- For interrupt callbacks, I wonder if drivers can register callbacks with the FlexIO when they claim resources? This way the FlexIO driver will know which driver callback to run based on which interrupt flags are set.

MikhailSiomin commented:
I'm currently very busy at work. At the end of this month I plan to resume development of this driver. I'm sorry for the delay.

manuargue commented on west.yml:
```diff
@@ -93,7 +93,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: d957472002a3758154f402e982d95ce90cf58e12
+      revision: 1b4be88e2e5d52d503adbccd2073ccc3fc874d0c
```
This is not resolved yet, you need to point the manifest to `revision: pull/205/head` as was already indicated

decsny commented on drivers/misc/mcux_flexio/mcux_flexio.h:
```diff
@@ -0,0 +1,30 @@
+#include <zephyr/drivers/spi.h>
+
+
+struct mcux_flexio_api {
```
> I don't see the value of adding this API structure. The MCUX FlexIO peripheral is NXP specific, and at least in the current form I don't see this API being implemented by other devices. Why not just use function declarations directly, and avoid dereferencing the API pointer?


@danieldegrasse  The answer to this question might be "multiple flexio on one platform" but I think it is a safe bet that would be very unlikely

danieldegrasse replied:
Still- you don't need an API structure. The implementations of the driver functions themselves won't look any different for the different FLEXIO instances, they are all the same IP block. So no need to abstract them behind an API struct.

mmahadevan108 commented:
@MikhailSiomin , this driver is very useful and we would like to make progress in this PR. Do you plan to get back to this? Let me know if you need help working on this PR. I would be happy to contribute/collaborate.

MikhailSiomin commented:
@mmahadevan108, Yes, I plan to return soon.

mmahadevan108 commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,481 @@
+	}
+}
+
+int spi_mcux_isr(const struct device *dev)
```
I would rename this function to `spi_mcux_flexio_isr`

mmahadevan108 commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,481 @@
+	return 0;
+}
+
+void spi_mcux_res(const struct device *dev, struct mcux_flexio_child_res *child_res)
```
Rename function to `spi_mcux_flexio_res`

MikhailSiomin commented:
I don't know how to do tests. Either describe for one board (using `platform_allow`), or describe all boards with `NXP FlexIO` (this will take a long time), or do no tests for this driver at all.

danieldegrasse commented:
@MikhailSiomin I think the best option is probably just using `platform_allow`. We could also use a twister tag like those seen here: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mimxrt1064_evk/mimxrt1064_evk.yaml#L17, but those are typically more generic than FlexIO is. Perhaps just add a testcase to the spi loopback sample that enables the MCUX FlexIO overlay and only has the RT1064 as an allowed platform:
```
spi_mcux_flexio_loopback:
    extra_args:
       - DTC_OVERLAY_FILE="overlay-mcux-flexio-spi.overlay"
     platform_allow:
       - mixmrt1064_evk
```

Also, thank you for updating the driver to use iterable sections. I think this approach makes more sense. Can you address compliance failures?

MikhailSiomin commented:
@danieldegrasse, notice the first commit. I had to make some changes for kinetis ke1xf series.

danieldegrasse commented on boards/arm/twr_ke18f/twr_ke18f.dts:
```diff
@@ -132,13 +132,6 @@
 		};
 	};
-
-	flexio: flexio@4005a000 {
```
I think we should mark the new (SOC level) flexio node as `status=okay` at the board level for the TWR-KE18F

danieldegrasse commented on drivers/misc/mcux_flexio/mcux_flexio.h:
```diff
@@ -0,0 +1,36 @@
+	uint8_t timer_count;
+};
+
+typedef int (*mcux_flexio_child_isr_t)(const struct device *dev);
```
A few comments on this callback signature:
- shouldn't we use the `dev` parameter to contain the FlexIO controller device itself? This would mirror the way other callbacks within Zephyr are designed
- Again following the pattern of other callbacks in Zephyr, maybe we should add a parameter `user_data`, which is defined within the `mcux_flexio_child` structure?

danieldegrasse commented on include/zephyr/linker/common-rom/common-rom-misc.ld:
```diff
@@ -39,6 +39,10 @@
 	ITERABLE_SECTION_ROM(zbus_channel_observation, 4)
 #endif /* CONFIG_ZBUS */
 
+#if defined(CONFIG_MCUX_FLEXIO)
```
I'm concerned about defining this `ITERABLE_SECTION_ROM` within the `common-rom-misc.ld` file, this location feels too generic. I'd rather see this macro placed in a snippet file of some kind within the `drivers/misc/mcux_flexio` folder, and included using `zephyr_linker_sources` within the FlexIO CMakeLists file, similar to the way Zephyr recommends applications use iterable sections: https://docs.zephyrproject.org/latest/kernel/iterable_sections/index.html#usage

danieldegrasse replied:
Another thought here- what are we gaining by using `STRUCT_SECTION_FOREACH` to initialize child FlexIO devices? Could we instead have each child FlexIO device define its own `struct mcux_flexio_child`, and call a function to "attach" to the parent FlexIO controller (something like `flexio_attach`?). That function could then allocate shifters and timers for the child device.

I suppose the reasoning here is that now if the FlexIO is overallocated than initialization will fail for all devices? If child devices requested allocation of FlexIO timers/shifters, than some FlexIO child devices could initialize, even if the FlexIO as a whole was overallocated in the application.

danieldegrasse commented on drivers/misc/mcux_flexio/mcux_flexio.h:
This file should be placed in the `include` folder, at a path like the following:
`include/zephyr/drivers/misc/mcux_flexio/mcux_flexio.h`

danieldegrasse commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,446 @@
+	spi_context_cs_control(&data->ctx, true);
+
+	mcux_flexio_lock(config->flexio_dev);
+	mcux_flexio_irq_disable(config->flexio_dev);
```
I'm unclear here- why do we need to disable the IRQ from the FlexIO before sending the first packet? I'm concerned when multiple FlexIO "devices" are defined, that disabling the interrupt could have adverse affects

MikhailSiomin replied:
@danieldegrasse, Because modification of general data occurs there: https://github.com/zephyrproject-rtos/zephyr/pull/53748/files#diff-d2bd4d8ff2692193e151b71d1ff172bfeca0ca52c197506a8c8548cabf05ff44R94

mmahadevan108 commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,446 @@
+		.dev = DEVICE_DT_INST_GET(n),				\
+		.isr = spi_mcux_flexio_isr,				\
+		.res = {						\
+			.shifter_index = flexio_spi_##n.shifterIndex,	\
```
These were previously derived from Device Tree. How are you getting these now?

MikhailSiomin replied:
@mmahadevan108, The FlexIO driver determines free resources and assigns them to the child: https://github.com/zephyrproject-rtos/zephyr/pull/53748/files#diff-1a74d72ae2e93b95139fdb5e22fbbac66251d0570a7f21c15571d7e191450fc6R160

danieldegrasse replied:
Sorry, I'm still a bit confused here- don't we want to take the address of these values here, IE `&flexsio_spi_##n.shifterIndex`?

MikhailSiomin replied:
@danieldegrasse, That's right, `shifterIndex` is an array, and its name is a pointer to the zero element. This will be the same as `&flexsio_spi_##n.shifterIndex[0]`. https://github.com/zephyrproject-rtos/hal_nxp/blob/d45b14c198d778658b7853b48378d2e132a6c4be/mcux/mcux-sdk/drivers/flexio/spi/fsl_flexio_spi.h#L121

danieldegrasse replied:
Understood- sorry, I should have checked this closer. All good from my side then

mmahadevan108 commented:
@MikhailSiomin, based on your availability, please do let us know if you need help to move further on the development of this driver further. We would be happy to help update this driver.

srkapil commented:
@MikhailSiomin Thanks a lot for doing this. I have a question regarding this PR. Is DMA support for Flex IO SPI part of this PR?

MikhailSiomin commented:
@srkapil, I didn't plan this, but I'll try to add DMA support for this driver.

harristomyx commented:
Hi, is there any updates on this PR?

MikhailSiomin commented:
@harristomyx, Hi. I'm gradually working on adding dma to this driver.

mmahadevan108 commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,447 @@
+		SPI_MCUX_FLEXIO_SPI_PINCTRL_INIT(n)			\
+	};								\
+									\
+	static const STRUCT_SECTION_ITERABLE(mcux_flexio_child, mcux_flexio_child_##n) = { \
```
Should we use a unique name such as `mcux_flexio_spi_child_##n` instead of `mcux_flexio_child_##n`

mmahadevan108 commented:
> @harristomyx, Hi. I'm gradually working on adding dma to this driver.

@MikhailSiomin, thank you for making the updates to the driver, overall the driver looks great. Can you help resolve the conflicts and also help resolve CI failures so that we can look to merging this PR.

Can we hold off on updating this PR with features such DMA. I request we make this addition as part of a separate PR.

mmahadevan108 commented:
@MikhailSiomin Thank you for your response. Could you help resolve the merge conflicts.

danieldegrasse commented on include/zephyr/drivers/misc/mcux_flexio/mcux_flexio.h:
```diff
@@ -0,0 +1,78 @@
+ * @brief Lock FlexIO mutex.
+ * @param dev Pointer to the device structure for the FlexIO driver instance
+ */
+void mcux_flexio_lock(const struct device *dev);
```
Just a note here- I'm not sure we should include `mcux_` as a prefix for these APIs. The driver does indeed rely on the MCUX HAL, but the API we are designing here is a consequence of the peripheral itself, not the HAL, and could probably be implemented without using the MCUX HAL if that was ever needed. Due to this, I think we should avoid the `mcux_` naming prefix here.

MikhailSiomin replied:
@danieldegrasse, then it is also worth removing the `mcux_` prefix from the path (i.e. `include/zephyr/drivers/misc/flexio/flexio.h`) and adding `dev_api` to the mcux_flexio driver so that future flexio drivers can implement functions from this header.

danieldegrasse replied:
If you'd like to do that, I'm fine with both- I guess the reasoning being that another driver might exist implementing this API without the MCUX HAL? Truthfully I think if such a driver existed it would just replace the existing driver, so we might not need the ability for multiple drivers to exist. I am fine if you'd prefer to add the `dev_api` though.

MikhailSiomin replied:
@danieldegrasse, flexio isn't really a driver in the Zephyr context, it's more of a driver for a driver that shouldn't provide access to users. So I suggest using `mcux_` prefixes in the file `include/zephyr/drivers/misc/flexio/flexio.h` and including this header in child drivers like this `#include "../misc/mcux_flexio/mcux_flexio.h"`

danieldegrasse replied:
I'm not sure I follow- will we have `mcux_` prefixes or not? I think users will absolutely access this driver, they need to do so in order to add support custom FLEXIO based peripheral drivers.

henrikbrixandersen replied:
MCUX sounds wrong as a prefix here, to be honest.

danieldegrasse replied:
I would strongly prefer we keep `mcux_` out of the naming scheme here. The API is not specific to MCUX SDK, it is enabling features of the FLEXIO IP. Can we place the file in `include/zephyr/drivers/misc/nxp_flexio/nxp_flexio.h`, and namespace functions with `nxp_flexio`?

danieldegrasse commented on drivers/spi/spi_mcux_flexio.c:
```diff
@@ -0,0 +1,431 @@
+	return 0;
+}
+
+static int spi_mcux_init(const struct device *dev)
```
Do we need to check if the parent FlexIO controller device is ready in this function, since that would indicate if the required shifters/timers for this driver were allocated successfully?

MikhailSiomin replied:
@danieldegrasse, There is no need to check the readiness of the parent device because the parent's initialization function does not affect resource management.

danieldegrasse commented on tests/drivers/spi/spi_loopback/testcase.yaml:
```diff
@@ -35,3 +35,10 @@ tests:
+    extra_configs:
+      - CONFIG_SPI_MCUX_FLEXIO=y CONFIG_SPI_ASYNC=y
+    filter: CONFIG_MCUX_FLEXIO
+    platform_allow: mimxrt1064_evk
```
@MikhailSiomin ping on this- do you see test failures when you remove the `platform_allow`?

MikhailSiomin replied:
@danieldegrasse, without `platform_allow`, some other tests break. I removed the commit with flexio driver test because I'm not sure that the complexity of adding this test is justified.

MikhailSiomin commented on drivers/misc/mcux_flexio/Kconfig:
```diff
@@ -0,0 +1,24 @@
+	depends on CLOCK_CONTROL
+	help
+	  Enable the FlexIO.
+	  FlexIO is not used separately.
```
@danieldegrasse, Do I understand correctly that you are suggesting to add additional information about where this configuration can be configured?

danieldegrasse replied:
That would be fine, yes. I was thinking something like the following (provided that everything in that help string is actually correct):

`Enable the FlexIO controller driver. This driver is not user-selectable, and should be enabled by other FlexIO drivers so that they can use it to share the resources of the FlexIO device`

decsny commented:
@MikhailSiomin 

NXP has strong interest in enabling the FlexIO in Zephyr. We have been very patient with you since you introduced this PR to let you take the authorship responsibility, but the pace that this is progressing is not ideal for our business. Would you like us to take over the responsibility to enable this, we can keep your name on the commits you have contributed thus far if that is important to you and we can finish the enablement much faster.

MikhailSiomin commented:
@decsny, OK I understood. I will devote the near future to finish this PR.

sumitbatra-nxp commented on drivers/misc/mcux_flexio/mcux_flexio.c:
```diff
@@ -0,0 +1,253 @@
+		mcux_fexio_map_timer_child_##n[MCUX_FLEXIO_TIMER_COUNT_MAX(n)] = {0}; \
+									\
+	static struct mcux_flexio_data mcux_flexio_data_##n = {		\
+		.map_shifter_child = mcux_fexio_map_shifter_child_##n, \
```
should be flexio instead fexio

mmahadevan108 commented:
@MikhailSiomin Can you please add the overlay files needed to test the SPI implementation.

MikhailSiomin commented:
@danieldegrasse, I couldn’t exclude `platform_allow` because then some tests break (dts is not generated correctly).

danieldegrasse commented on drivers/spi/CMakeLists.txt:
```diff
@@ -11,6 +11,7 @@ zephyr_library_sources_ifdef(CONFIG_SPI_EMUL		spi_emul.c)
 zephyr_library_sources_ifdef(CONFIG_SPI_STM32		spi_ll_stm32.c)
 zephyr_library_sources_ifdef(CONFIG_SPI_MCUX_DSPI	spi_mcux_dspi.c)
 zephyr_library_sources_ifdef(CONFIG_SPI_MCUX_FLEXCOMM	spi_mcux_flexcomm.c)
+zephyr_library_sources_ifdef(CONFIG_SPI_MCUX_FLEXIO spi_mcux_flexio.c)
```
Nit- align indentation

danieldegrasse commented on dts/bindings/spi/nxp,flexio-spi.yaml:
```diff
@@ -0,0 +1,32 @@
+    description: |
+      Pin select for clock.
+
+child-binding:
```
Do we need `child-binding` here? Other SPI controllers do not have it.

danieldegrasse reviewed (commented):
Thanks for all the work here- besides my comment about the `mcux_flexio_*` header and function naming, generally think this PR is in a good state.

danieldegrasse commented:
> @danieldegrasse, I couldn’t exclude platform_allow because then some tests break (dts is not generated correctly).

Which tests specifically? I'd be happy to take a look to see if we can avoid `platform-allow`

mmahadevan108 commented on drivers/misc/mcux_flexio/mcux_flexio.c:
```diff
@@ -0,0 +1,253 @@
+
+				if (child != NULL) {
+					child->isr(child->user_data);
+				}
```
We should check if the isr field is not NULL before calling it. Not every driver might be interested in the ISR as they may use DMA.

mmahadevan108 commented on tests/drivers/spi/spi_loopback/boards/mimxrt1064_evk.overlay:
```diff
@@ -5,14 +5,86 @@
+			spi-max-frequency = <4000000>;
+		};
+	};
+};
```
Why are the above lines added to this overlay file. Should they be in the `flexio-overlay` file

MikhailSiomin replied:
@mmahadevan108, I made it so that the test could be run on the board. If a successful build is enough for testing, then in the `flexio-overlay` file it possible configure only `flexio1` without pinctrl.

mmahadevan108 replied:
@MikhailSiomin, can we move the flexio changes out of`boards/mimxrt1064_evk.overlay` file and move them to `overlay-mcux-flexio-spi.overlay`. We can use the build command:` west build -p -b mimxrt1064_evk zephyr/tests/drivers/spi/spi_loopback -T drivers.spi.mcux_flexio_spi.loopback` to pull the right overlay file when testing for FlexIO.
With the current changes, FlexIO gets enabled when building for the default spi_loopback test which is incorrect.

mmahadevan108 commented on drivers/misc/mcux_flexio/mcux_flexio.h:
```diff
@@ -0,0 +1,88 @@
+	uint8_t shifter_count;
+	uint8_t *timer_index;
+	uint8_t timer_count;
+};
```
Is this driver not managing the pins anymore?

mmahadevan108 replied:
The previous version of your implementation had an entry for the pins. I guess the expectation is that FlexIO pins are managed by the individual drivers based on the board dts file. Am I understanding this correctly?

MikhailSiomin replied:
@mmahadevan108, Yes, the selection and configuration of pins is determined in dts using `sdo-pin`, `sdi-pin`, `sck-pin`, `cs-gpios` and `pinctrl`. Any shifter or timer can be mixed with any pin, so for convenience, their assignment is handled by the parent driver.

MikhailSiomin commented:
@danieldegrasse, The `overlay file` from `extra_args` is also overlaid on platforms where there is no flexio and/or lpspi and/or pinctrl. If you run tests without `platform_allow` as follows, then `nativ_sim` will not build: `python .\scripts\twister --short-build-path -T tests/drivers/spi/ -s tests/drivers/spi/spi_loopback/drivers.spi.mcux_flexio_spi.loopback`

mmahadevan108 commented on drivers/misc/mcux_flexio/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+
+zephyr_library()
+
+zephyr_library_sources(mcux_flexio.c)
```
Can we add `zephyr_include_directories(.)` to this file. That way when including the header file, we don't have to include the complete path.

decsny replied:
I don't think that's the way to do this; the header should go in the include directory in the first place if it is intended to be a public header

mmahadevan108 replied:
That sounds like a better suggestion. We could move the .h file to the include folder.

MikhailSiomin replied:
I will do it.

danieldegrasse commented on drivers/clock_control/clock_control_mcux_ccm.c:
```diff
@@ -327,6 +327,53 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
+		*rate = source_clk_freq / (CLOCK_GetDiv(kCLOCK_Flexio1PreDiv) + 1)
+					/ (CLOCK_GetDiv(kCLOCK_Flexio1Div) + 1);
+	} break;
+#endif
```
Shouldn't the `#endif` here be after the case for `IMX_CCM_FLEXIO2_3_CLK`? Also, should we check if `DT_NODE_HAS_STATUS(DT_NODELABEL(flexio1), okay)` for the first case definition here?

MikhailSiomin replied:
@danieldegrasse, I agree that there should be a check for `DT_NODE_HAS_STATUS(DT_NODELABEL(flexio1), okay)`, because `flexio1` is not always enabled. But `#endif`, it seems to me better not to change it, as is done in lines `214` and `221` of the same file.

danieldegrasse replied:
Yeah, you are right- using `&&CONFIG_MCUX_FLEXIO` with each conditional is best here

bia-bonobo commented on dts/arm/nxp/nxp_rt1060.dtsi:
Could you also add the flexio2 and flexio3 to the nxp_rt1040.dtsi? The 1040 has also flexio1, flexio2 and flexio3.

danieldegrasse reviewed (dismissed):
Thanks for all the work here- I think this PR is good to go once the merge window opens back up. We will need to merge the HAL side PR once this one is approved though

mmahadevan108 commented on drivers/misc/mcux_flexio/Kconfig:
```diff
@@ -0,0 +1,26 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config MCUX_FLEXIO
+	bool
```
Is this missing a line `default y`

mmahadevan108 commented on drivers/spi/Kconfig.mcux_flexio:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SPI_MCUX_FLEXIO
+	bool "MCUX FlexIO SPI driver"
```
Is this missing `default y`

MikhailSiomin replied:
@danieldegrasse, Is it worth making this amendment to this PR?

danieldegrasse replied:
Yes- thank you for adding it

mmahadevan108 commented on tests/drivers/spi/spi_loopback/testcase.yaml:
```diff
@@ -158,3 +158,9 @@ tests:
+  drivers.spi.mcux_flexio_spi.loopback:
+    extra_args: EXTRA_DTC_OVERLAY_FILE="overlay-mcux-flexio-spi.overlay"
+    extra_configs:
+      - CONFIG_SPI_MCUX_FLEXIO=y
```
Not needed if we `default y` the driver Kconfig

MikhailSiomin committed:
dts: nxp: Added flexio support for i.MX rt10xx and kinetis ke1xf series

Added description of flexio peripherals for i.MX rt10xx
and kinetis ke1xf series.

Signed-off-by: Mikhail Siomin <victorovich.01@mail.ru>

MikhailSiomin committed:
clock: nxp_imx: Added clock control support for FlexIO

Added clock control support for flexIO for i.MX series.

Signed-off-by: Mikhail Siomin <victorovich.01@mail.ru>

danieldegrasse commented:
@MikhailSiomin can you take a look at the twister failures?

MikhailSiomin commented:
@danieldegrasse, It seems to me that failed tests are not related to changes in the PR.

MikhailSiomin commented:
Resolved merge conflicts.

mmahadevan108 commented:
@MikhailSiomin , thank your contribution. The HAL PR has been merged. Can you please update the Zephyr side.

MikhailSiomin committed:
drivers: mcux: flexio: Added generic MCUX FlexIO driver

Added FlexIO driver that distributes hardware resources
between interfaces using them.

Signed-off-by: Mikhail Siomin <victorovich.01@mail.ru>

MikhailSiomin committed:
drivers: mcux: flexio: Added MCUX FlexIO SPI driver

Added SPI driver using FlexIO.

Signed-off-by: Mikhail Siomin <victorovich.01@mail.ru>

MikhailSiomin committed:
tests: spi: spi_loopback: Added MCUX FlexIO SPI test

Added FlexIO SPI test. Overlay in the tests for mimxrt1064_evk
board only.

Signed-off-by: Mikhail Siomin <victorovich.01@mail.ru>

MikhailSiomin commented:
Thank you all for the work done on the PR.

mmahadevan108 commented:
@MikhailSiomin , I am trying to test this on the 1060EVK board. Can you please share with me the connections you had to make on the board in order to run this test.

MikhailSiomin commented:
@mmahadevan108, To make spi tests in `loopback` mode successful, i put a `jumper` between `sdi` and `sdo` pins for each flexio_spi.

mmahadevan108 commented:
@MikhailSiomin thank you for your response.
